### PR TITLE
DATAUP-616: ensure job messages are not lost

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

Whilst prepping the job comm channel for the new message type, I fixed the issue where messages sent before the comm channel to the backend is opened get lost in the eternal void of damnation, and added some appropriately hellish tests.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-616
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative, checking to see if there are any errors about messages not being sent due to the job comm channel not being opened, and feeling a rush of relief in knowing that one day, those messages will be sent. Phew!

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/buildTools/loadAppWidgets.js [hunk 1]
```diff
@@ -84,8 +84,6 @@ define([
     'common/jupyter',
     'common/lang',
     'common/messages',
-    'common/microBus',
-    'common/miniBus',
```
Comment (ialarmedalien) [inline kbase-extension/static/buildTools/loadAppWidgets.js:88]: these two files are not used ==> delete!

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 2]
```diff
@@ -148,6 +148,9 @@ define([
                 : () => {
                       /* no op */
                   };
+
+            this.messageQueue = [];
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:152]: a seemingly-innocent message queue

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 3]
```diff
@@ -178,76 +181,99 @@ define([
...
-                // or try to init comm channel here
-                if (!this.comm) {
-                    console.error(
-                        'Comm channel not initialized, not sending message.',
-                        msgType,
-                        message
-                    );
-                    reject(new Error('Comm channel not initialized, not sending message.'));
-                }
+        _transformMessage(rawMessage) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:199]: separate out the transformation of the message into a different sub

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 4]
```diff
@@ -178,76 +181,99 @@ define([
...
+         * If no arguments are supplied to sendCommMessage, the stored messages (if any)
+         * will be sent.
+         *
+         * @param {string} msgType - message type; will be one of the
+         *                 keys in the requestTranslation object (optional)
+         *
+         * @param {object} msgData - additional parameters for the request,
+         *                 such as jobId or jobIdList, or an 'options' object (optional)
+         */
+        sendCommMessage(msgType, msgData) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:246]: This method signature is useful for testing, so I left it in; `_sendCommMessages` takes messages from the `messageQueue` and sends them to the backend.

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 5]
```diff
@@ -545,21 +571,21 @@ define([
...
+                                    commSemaphore.set('comm', 'error');
+                                    reject(
+                                        new Error(reply.content.name + ': ' + reply.content.evalue)
+                                    );
                                 },
                             },
                         };
                         Jupyter.notebook.kernel.execute(_this.getJobInitCode(), callbacks);
+                        if (_this.messageQueue.length) {
+                            _this.sendCommMessage();
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:587]: after initing the comm channel, send any queued messages

File: kbase-extension/static/kbase/js/common/jobs.js [hunk 6]
```diff
@@ -25,7 +25,7 @@ define(['common/errorDisplay', 'common/format', 'common/html', 'common/ui'], (
             'This job was not found, or may not have been registered with this narrative.',
             'You will not be able to inspect the job status or view the job log.',
         ],
-        jobStatusUnknown = ['Determining job state...'],
+        jobStatusUnknown = ['Awaiting job data...'],
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobs.js:28]: changed this to be slightly more accurate about what is going on

File: kbase-extension/static/kbase/js/common/monoBus.js [hunk 7]
```diff
@@ -91,7 +91,7 @@ define(['uuid', 'bluebird', 'underscore', './lang'], (Uuid, Promise, _, lang) =>
         function ensureChannel(name) {
             if (!channels[name]) {
                 warn(`Channel implicitly created: ${name}`);
-                makeChannel({ name: name });
+                makeChannel({ name });
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/monoBus.js:94]: all changes to this file are just `{ this: this }` => `{ this }`

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 8]
```diff
@@ -178,76 +181,99 @@ define([
...
+        _sendCommMessages() {
+            let topMessage;
+            return Promise.try(() => {
+                if (!this.comm) {
+                    // TODO: try to init comm channel here
+                    throw new Error('Comm channel not initialized, not sending message.');
+                }
+                while (this.messageQueue.length) {
+                    topMessage = this.messageQueue.shift();
+                    this.comm.send(this._transformMessage(topMessage));
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:271]: separate out all the transformation stuff to simplify this method

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 9]
```diff
@@ -178,76 +181,99 @@ define([
...
+         *
+         * @param {string} msgType - message type; will be one of the
+         *                 keys in the requestTranslation object (optional)
+         *
+         * @param {object} msgData - additional parameters for the request,
+         *                 such as jobId or jobIdList, or an 'options' object (optional)
+         */
+        sendCommMessage(msgType, msgData) {
+            if (msgType && msgData) {
+                this.messageQueue.push({ msgType, msgData });
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:248]: if there is a new message, push it on to the messageQueue

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 10]
```diff
@@ -178,76 +181,99 @@ define([
...
+         */
+
+        _sendCommMessages() {
+            let topMessage;
+            return Promise.try(() => {
+                if (!this.comm) {
+                    // TODO: try to init comm channel here
+                    throw new Error('Comm channel not initialized, not sending message.');
+                }
+                while (this.messageQueue.length) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:269]: FIFO iteration through the message queue

File: kbase-extension/static/kbase/js/common/jobCommChannel.js [hunk 11]
```diff
@@ -545,21 +571,21 @@ define([
...
-                                        console.error('ERROR executing jobInit', reply);
-                                        commSemaphore.set('comm', 'error');
-                                        reject(
-                                            new Error(
-                                                reply.content.name + ':' + reply.content.evalue
-                                            )
-                                        );
-                                    } else {
-                                        resolve();
+                                    if (!reply.content.error) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/jobCommChannel.js:574]: no content change, just return early if there is no error in the `reply`

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 12]
```diff
@@ -243,19 +447,26 @@ define([
                     .initCommChannel()
                     .then(() => {
                         expect(comm.comm).not.toBeNull();
+                        spyOn(comm, 'sendCommMessage').and.callFake((...args) => {
+                            return comm.sendCommMessage.and.originalFn.call(comm, ...args);
+                        });
```
Comment (ialarmedalien) [inline test/unit/spec/common/jobCommChannel-spec.js:452]: add in an extra spy -- this was mostly for symmetry with the previous tests involving the stored messages

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 13]
```diff
@@ -243,19 +447,26 @@ define([
...
                     })
-                    .then((args) => {
+                    .then(() => {
                         expect(comm.comm.send).toHaveBeenCalled();
-                        expect(args[0]).toEqual(
-                            Object.assign(testCase.expected, { target_name: 'KBaseJobs' })
-                        );
+                        expect(comm.comm.send.calls.allArgs()).toEqual([
+                            [Object.assign(testCase.expected, { target_name: 'KBaseJobs' })],
+                        ]);
```
Comment (ialarmedalien) [inline test/unit/spec/common/jobCommChannel-spec.js:465]: Pretty much the same as the old test, just ensures that there has only been one call.

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 14]
```diff
@@ -129,24 +132,225 @@ define([
...
+                            resolve();
+                        });
+                        // emit the last ping
+                        testBus.emit(...pingArgs[2]);
+                    }).then(() => {
+                        expect(comm.sendCommMessage).toHaveBeenCalledTimes(pingArgs.length + 1);
+                        expect(comm.sendCommMessage.calls.allArgs()).toEqual([
+                            pingArgs[0],
+                            pingArgs[1],
+                            [],
```
Comment (ialarmedalien) [inline test/unit/spec/common/jobCommChannel-spec.js:262]: this is from when `comm.sendCommMessage()` gets called at the end of the `comm.initCommChannel` method.

File: test/unit/spec/common/jobCommChannel-spec.js [hunk 15]
```diff
@@ -129,24 +132,225 @@ define([
...
         it('Should error properly when trying to send a comm with an uninited channel', async () => {
             const comm = new JobCommChannel();
+            expect(comm.comm).toBeUndefined();
             await expectAsync(comm.sendCommMessage('some_msg', {})).toBeRejectedWithError(
-                /ERROR sending comm message: /
+                'ERROR sending comm message: ' + comm.ERROR_COMM_CHANNEL_NOT_INIT
             );
         });
 
+        it('should not lose messages if the comm channel is not inited', () => {
```
Comment (briehl) [inline test/unit/spec/common/jobCommChannel-spec.js:155]: These are some really twisty tests. Appropriately hellish, as you said! :fire:

**Review by briehl — APPROVED** [review]
Looks good. Tricky, but looks good, I think.

**Pull request merged**